codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -1036,11 +1043,11 @@ private void checkNewStreamAllowed(int streamId, State state) throws Http2Except
             }
             // This check must be after all id validated checks, but before the max streams check because it may be
             // recoverable to some degree for handling frames which can be sent on closed streams.
-            if (streamId < nextStreamId) {
+            if (streamId < nextExpectedStreamId) {
                 throw closedStreamError(PROTOCOL_ERROR, "Request stream %d is behind the next expected stream %d",
-                        streamId, nextStreamId);
+                        streamId, nextExpectedStreamId);
             }
-            if (isExhausted()) {
+            if (nextExpectedStreamId <= 0) {
```

Comment:
I think we would lose the ability to check the stream error first then.

```java
// This check must be after all id validated checks, but before the max streams check because it may be
// recoverable to some degree for handling frames which can be sent on closed streams.
```